### ISSUE TITLE
add a property of 'supplier price' and a 'split' value

### ISSUE THREAD

**Issue by carchrae** [foodcoops/foodsoft#474]
our co-op, and i think many others, order cases of items from our suppliers.  so, a case of 100 apples might be $50 for 100 apples.  while we show the price to our members as $0.50 per apple and unit quantity (UQ) of 100, the supplier price is $50 and this is what must be communicated to the supplier when ordering.

i have written a bunch of scripts that convert the supplier price into a price per item, but this isn't ideal.  i would prefer to have the original price stored and derive the co-op price from the original price / UQ.

here are some examples of the problems right now:

1. the rounding problem.  so that the co-op does not lose money, i round the price up.  so, if there are 66 pears in a box that costs $50, then price per pear is $0.7575757575 - or rather $0.76.  now, when i compute the supplier price from my derived price of $0.76, the case cost is listed as 66 UQ * 0.76 = $50.16.  the supplier may be confused at the price discrepancy.  (and sure i can see how the coop can slowly get rich!). 

2. if the price of the box changes, it requires more calculating to determine the food coop price.  for example, say some of the fruit is spoiled/bruised, and the supplier says the box will cost $30, then we cannot update the case price, but need to do more calculating.  

3. if the amount in the box is different, then we need to determine the updated cost based on the new quantity.

i propose that we add a 'case price' to the article model.  this price can then be used when producing the ordering pdf/fax to the supplier.  it can also be used when updating the costs of articles in a particular order.

**Comment by wvengen** [comment]
Wow, good practical cases that I recognize from other fresh-based foodcoops.
(1) Adding a case or box price would be great to solve rounding errors. Need to think how to present this clearly without clutter..
(2) and (3) perhaps this may become part of the receive screen, where differences in what was ordered and what was received are handled. Now only member amounts are taken into account, but this could also take article price into account. Since the `order_article` already has its own `article_price`, we can easily assign a new price to it (without affecting the `article`'s price). Just need to make sure that it's clear from the UI why there may be different prices (e.g. message explaining or whatever).

**Comment by carchrae** [comment]
on 1) i'll try a few things and see what i can come up with.  because i know a lot of logic depends on the current setup, i will try and keep the changes minimal.  

thanks for the reminder about `article_price` - i will have a deeper look at the code and see how the received discrepancies are handled in the received/settling order pages.  

for example, i like that you can put fractions in the amount received by a co-op member.  but i would like it if i could edit the unit in that screen too - for example, 1 qty X 250g of dried mango, actual weight member receives is 272g.  i can update it as 1.088 units (250*1.088=272g).  i am inclined to make these kind of changes outside the model, for example, the UI allows you to computes the unit by entering a price (instead of using a calculator, like i just did)

**Comment by wvengen** [comment]
The unit is current part of the `article` - so changing that is not recommended (I've wondered whether that should become part of `article_price` or not). This can indeed be a UI-thing (the foodcoop-adam fork allows editing grams, as an example.

![image](https://cloud.githubusercontent.com/assets/503804/25673796/e2173fac-3038-11e7-956b-fcd7367a8e2e.png)
![image](https://cloud.githubusercontent.com/assets/503804/25673815/f63b4672-3038-11e7-9e6f-85caf053da28.png)

**Comment by Nalaxon** [comment]
Hey there :)
Are there!
Any news about this?
We would highly appreciate such feature

@carchrae
How you deal with shipping costs? Would you put that in the box- price as well?
Currently we looking for a solution for that.

Thanks in advance!

**Comment by wvengen** [comment]
Hi, thanks for the question! My focus is first on getting [foodsoft-shop](https://github.com/foodcoops/foodsoft-shop) done, then using the mechanisms decided on there to implement a toggle between piece/unit (whether to do that in Ruby or in JS+React, for example; there still is a little discussion on what way to go).
But then again, I don't find _that_ many commits related to this, maybe https://github.com/foodcoop-adam/foodsoft/commit/8ef46ebcd46f8f5f4370a50b6df3974803e7396b would be enough (too long ago to remember).

**Comment by carchrae** [comment]
you can check my fork. it supports supplier price. i would really like to find time to merge it back to master, i have many features and speed improvements to contribute back.

for shipping cost, it is a hack but you can use deposit for a fixed cost, or tax as s percent cost.  i do need to make a new feature to split a shipping cost among orders. it is cumbersome to do that now. please make a new issue for that.

also agree with @wvengen.  many features should be ui customization, not server changes. supplier price however does require an extra attribute.

**Comment by mischkl** [comment]
The feature for a weight article would be very helpful and desperately needed!

We are in the process of switching to FoodSoft and are confronted with the problem that there is no way to enter sub-1kg amounts for articles that are priced by the kg.

So far we're worked around this by setting the unit to 100g, but then the price has to be divided by 10 and gets rounded up or down in comparison to the kg-price - not good! Not only does the supplier receive an order form with prices and units that are not quite correct, but people who pick up their orders can't rely on the price as listed in FoodSoft since it always has a rounding error.

For reference I've also commented on #422, either solution would work for us. The sooner this feature can be provided the better! :)

**Comment by mischkl** [comment]
Also I see there are issues going back to 2014 related to this. #101 #209 #296

**Comment by carchrae** [comment]
@mischkl - you can set the unit to 100g *and* the unit quantity to 10 - then the supplier gets an order by a quantity of 1kg.  on the rounding of prices, yes, you need to round up to stop your co-op from losing money slowly, and then the price to the supplier is wrong.  

however, i agree that it isn't that clear to the supplier.  i customised our supplier PDF to try to make it more clear to the supplier.  i also added a property to store the supplier price.  my version is a fork and is now out of date.  i would like to merge the changes, but i have been very overloaded with my day job(s) (and then helping my coop is also a lot of work).

![image](https://user-images.githubusercontent.com/959825/35767580-aec9fb0a-08a3-11e8-808c-97c639c18260.png)

you are welcome to try my version, too.  https://github.com/carchrae/foodsoft  - if you are a programmer, it would be great to have some help merging some of the changes back in to the main project.

**Comment by mischkl** [comment]
@carchrae the problem is not that the supplier requires the units in whole kgs, we have been ordering in tenths of a kilogram for a long time. The main problem is that the price is rounded and therefore even the workaround of using smaller units is not a real solution.

**Comment by mischkl** [comment]
I am a programmer but also don't have much free time. I'd like to contribute but am honestly wondering what the solution is. It seems like with the number of partial solutions already hanging around in various forks for months or years, it should just be a matter of merging one. In my experience though this is usually up to the maintainers and not newbies to a project such as myself.

**Comment by carchrae** [comment]
yes, that was a problem for me too.  that is why in my version i store the supplier price separately.  if you don't want to use my version, another (painful) workaround is to edit all the prices before sending the order off.  

there is another annoying rounding bug when it comes to received items.  a case can often be a little less than it should be, eg, 200g less.  if 1kg was ordered, that would be nicely entered as 'received 0.8' - which works fine.  however, it doesn't always end up so nice, eg, 3 kg and you receive 200g less, then you have received '0.93333333333333333....' - it makes it harder to balance the order for accounting.  minor, but adds up

**Comment by carchrae** [comment]
> I am a programmer but also don't have much free time. I'd like to contribute but am honestly wondering what the solution is. It seems like with the number of partial solutions already hanging around in various forks for months or years, it should just be a matter of merging one. In my experience though this is usually up to the maintainers and not newbies to a project such as myself.

none of us have much free time.  @wvengen has been amazing in terms of his devotion and gifting his time to this project, which @bennibu also gave a lot of time/effort to.  

most open source projects have programmers using them who are employed by companies.  i doubt any of us are employed by our co-ops - and indeed, it would probably make the co-op bust if they had to pay.  so, i am afraid we are all DIY here. 

however. i can say this is 10000x times better than the horrible ASP.net VB code that our coop had been maintaining for 15 years.  there are still some big problems with this project too in terms of code/design.  

so i say to you, give a little back, it is appreciated.  or fork your own and then the next guy will make the same complaint as you.  ;)  (i am truly sorry i didn't merge all my changes back in)

**Comment by carchrae** [comment]
>  (i am truly sorry i didn't merge all my changes back in)

*yet*

**Comment by carchrae** [comment]
there are also some paid solutions; many of them ancient.  i am not sure how responsive they are to feature requests, but this is actually a direction some co-ops may want to consider, eg https://www.managemy.coop/

i didn't choose that route because our co-op had a specific way of working that i wanted to support.

**Comment by twothreenine** [comment]
I think the rounding discrepancy should be in a tolerable range now thanks to 02853a5 (Article versioning, extended units and alternative article sharing - #1073)

**Comment by lentschi** [comment]
> I think the rounding discrepancy should be in a tolerable range now thanks to https://github.com/foodcoops/foodsoft/commit/02853a5268742d0641f77d23c4c9f81b83d933ee (Article versioning, extended units and alternative article sharing - https://github.com/foodcoops/foodsoft/pull/1073)

👍 Yes, but there _are_ some remaining issues. Discussions can continue here, if required: https://github.com/foodcoops/foodsoft/issues/1092#issuecomment-2646427578

**Comment by carchrae** [comment]
in my view, you should show the rounding errors.  i've modified our balancing page to do that.  (it isn't perfect - most of our items are tax free, and there are still some rounding issues there)

the most important thing is that someone can see the order properly balance ($100 paid to supplier, $100 paid from members + $0.xx paid from members in rounding errors)

![Image](https://github.com/user-attachments/assets/2cdb2965-dfc8-43c4-8a43-8f27d30d035e)

**Comment by carchrae** [comment]
if you're curious, all this balancing work is in my fork: https://github.com/carchrae/foodsoft

i would have loved to submit it back, but most of it is not perfect and i was short on time

**Comment by carchrae** [comment]
here is an example of a single item, counting the rounding error.

important to note that the rounding error relates not to low-level 64/32 bit floating point errors, but rather rounding to the lowest amount of currency (aka a cent).

<img width="972" alt="Image" src="https://github.com/user-attachments/assets/09f325c1-d386-4fc2-ad02-406d3c0eaec7" />

**Comment by twothreenine** [comment]
I just tested your original example:

> if there are 66 pears in a box that costs $50, then price per pear is $0.7575757575 - or rather $0.76. now, when i compute the supplier price from my derived price of $0.76, the case cost is listed as 66 UQ * 0.76 = $50.16. the supplier may be confused at the price discrepancy. (and sure i can see how the coop can slowly get rich!).

With the new article units system, this can be entered like this:

![Image](https://github.com/user-attachments/assets/68bd18eb-2193-4cc9-ab1d-7647eecfb5ab)

When 3 ordergroups ordered 22 pears each, this is what it looks like in the accounting menu:
(I guess we should change it so it displays 22 pieces each instead of 0.333 parcels, so you don't have to use the conversion popup in this common use case, but that's just a UI issue)

![Image](https://github.com/user-attachments/assets/40f1530f-bc4f-4728-96fb-ccfd80314c10)

**Comment by carchrae** [comment]
your example is why i don't use the co-op price and why i store the
supplier price separately.  it is impossible in any case with a rounding
error for the unit_price * unit_quantity to equal the supplier price.  (in
my screenshots, you can see the supplier price is shown and compared
against the co-op price)

**Comment by twothreenine** [comment]
I think we're talking past each other. 🤔
You can see in my screenshots that with the new units system, there is no such rounding error.
You store the original supplier price in Foodsoft -- 50.00 € per parcel. The price per pear will be calculated much more precisely and is not rounded to 2 decimals. The total costs per ordergroup (16.67 each) are displayed rounded to 2 decimals but stored with more precision. That's why in sum, it's 50.00, equal to the original supplier price.

**Comment by twothreenine** [comment]
> round when you charge an ordergroup

Oh, I thought the transaction gets saved with more precision than 2 decimals, but it is `t.decimal "amount", precision: 8, scale: 2`.
So in this case, the group sum would be 3 * 16.67 = 50.01 and the food coop would make a profit of 0.01 €. (Currently the group sum gets displayed incorrectly as 50.00, but my PR #1082 (which I'll rebase now) will fix that.)
Perhaps we should increase the precision of transactions, too.

**Comment by carchrae** [comment]
i suppose it depends on how accurate you'd like to be.  higher precision is a nice, and more fair, fix to consider "fractional cents" in all the money management parts.

in practice, i don't think most members care about a few cents - the biggest headache for me was ensuring the order was properly balanced.  if supplier invoice doesn't match what members are charged, you can't be really sure (and even if does match perfectly, it could still be off - but much less likely).  i think your precision improvements will ensure it does match (on the balancing page) without resorting to any rounding.

thanks again!

**Comment by carchrae** [comment]
i'll have to have a closer look at your work when i get a moment.  in my branch, it shows the 0.10 cent type error in the place where it shows $1.61 below.

i also added a motivational [hey-ya](https://youtu.be/ZZ5LpwO-An4) gif when people balance it perfectly.  you may laugh, but it works.

![Image](https://github.com/user-attachments/assets/c62ea82d-31a5-4d28-963b-3a0b0d4bc8e2)

i would love to get my branch back to sync with the current work.  there are numerous things i've added to improve it (like emailing updates when prices change, when the final charges when the order is settled, etc) and the next thing i want to add is an email that says  'order is closing in X hours and these items need filling'

anyway, lots of ideas, not enough time.  story of my life.

**Comment by twothreenine** [comment]
Ok but that's basically the same field as _FC surplus_ that has been upstream for several years now, right?
It shows the balance between _ordergroups sum_ (with/without _extra charge_ aka _margin_) and _invoice amount_ (_adjusted for refund_ aka deposit exluded).
But, I noticed that ordergroups sum has been calculated incorrectly and tried to fix that in my PR. With this example I actually just noticed that I should fix the calculation _groups without charge_ as well, else you get this:

(Same `Pear` example as above. _FC surplus without extra charge_ should be 0.01 as well.)

![Image](https://github.com/user-attachments/assets/ca08a285-298d-4588-ad9c-de9001b836f0)

**Comment by carchrae** [comment]
no, our group doesn't use the surplus % charge.  the $1.61 in my screenshot is the total of rounding up errors.  without splitting out the $1.61 it would show $1292.90 + 1.16 = 1294.21 as the members paid amount.

the net charge is the amount that shows on the supplier's invoice (well, with any adjustments from refunds/etc).

to avoid people being sloppy it was important that the supplier's amount was equal to the amount without the rounding errors.  then we could get to $0 difference.

**Comment by twothreenine** [comment]
> the $1.61 in my screenshot is the total of rounding up errors

And those rounding errors are the discrepancy between supplier prices and group order unit prices, like rounding 50/66 = 0.7575... to 0.76 -- or something else? 

The discrepancy between supplier prices and group order unit prices should be below one cent with the new unit handling,
The discrepancy between ordergroups charges and invoice amount will be displayed in the FC surplus field, with my PR.
I don't think there's any discrepancy left, or am I wrong?

**Comment by carchrae** [comment]
> And those rounding errors are the discrepancy between supplier prices and group order unit prices, like rounding 50/66 = 0.7575... to 0.76 -- or something else as well?

yeah, specifically this:

https://github.com/carchrae/foodsoft/blob/d3dbf0a8d7f55c922a258e55ad47f55416c6e00d/app/models/order_article.rb#L111

so i think your precision fix will likely handle this.  the $1.61 shown in the screenshot is the sum of these errors across all OrderArticle records in that order.  however, unless you keep the order group balances in higher precision, you'll have to at some point do some rounding.  i like the idea of keeping the balances in higher precision though - a food co-op should not stoop to stealing money from members through rounding errors! :D  

i didn't consider using FC surplus at all, and although technically a rounding error in favour of the group is the same thing, it might be a bit confusing to see FC surplus amounts when you have 0% overhead.

**Comment by lentschi** [comment]
Late to the discussion and I might be missing some resolutions you have come to.
Just about this point:

> Perhaps we should increase the precision of transactions, too.

I think the idea was to have transaction amounts mirror the actual behavior of bank transactions: You cannot move 0.001 € between accounts. If the precision is different, the resulting account states will also deviate over time.

Also, no matter how high we set the precision, the precision will never be _perfect_. (See our general discussion [here](https://github.com/foodcoops/foodsoft/issues/1092#issuecomment-2646427574) - this is about amounts, but the issue is very similar.)

**Comment by carchrae** [comment]
yes, it will never be perfect, but as you say, for all the amounts inside the system, we could do better than the banks! :D
